hxprintlinks.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXPRINTLINKS" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxprintlinks \- add a numbered list of links at the end of an HTML file
.SH SYNOPSIS
.B hxprintlinks
.RB "[\| " \-b
.IR base " \|]"
.I file-or-URL
.SH DESCRIPTION
.B hxprintlinks
adds a numbered list of all URLs (links) found in the file to the end of
the HTML file.
.SH OPTIONS
The following options are supported:
.TP
.BI \-b " base"
Make the URLs in the list absolute, using
.I base
as the base URL.
.SH OPERANDS
The following operand is supported:
.TP
.I file-or-URL
The file to work on. If the argument is absent, or if it is - (a
single dash), the document is read from the standard input.
.SH "SEE ALSO"
.BR hxcopy (1)
.SH BUGS
.LP
The option
.B \-b
is probably of limited use. It makes the links in the generated list
absolute, but doesn't change the other links. For a more flexible
solution, use
.BR hxcopy (1)
on the output of
.BR hxprintlinks .
E.g., instead of
.d
hxprintlinks -b http://example.org/base/ myfile.html
.e
use
.d
hxprintlinks myfile.html | hxcopy -i http://example.org/base -o .
.e


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxprintlinks.c
/*
 * Add a numbered list of links at the end of an HTML file
 *
 * Copyright © 2001-2015 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/2002/copyright-software-20021231
 *
 * Created 23 Jan 2015 (based on a Perl version from 1 Feb 2001)
 * Bert Bos <bert@w3.org>
 */
#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#ifdef HAVE_STRING_H
#  include <string.h>
#elif HAVE_STRINGS_H
#  include <strings.h>
#endif
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include "types.e"
#include "dict.e"
#include "openurl.e"
#include "errexit.e"
#include "heap.e"
#include "html.e"
#include "scan.e"
#include "url.e"


static conststring attname[] = {	/* Attributes that contain URLs: */
  "src", "href", "data", "longdesc", "cite", "action", "profile",
  "background", "usemap", "classid", "codebase"};
static pairlist list = NULL;		/* Stored list of URLs */
static bool has_error = false;		/* Parsing errors occurred */
static conststring base = NULL;		/* Make URLs relative to this base */


/* pairlist_push -- insert a name/value pair at the start of a list */
static void pairlist_push(pairlist *p, const conststring name, const conststring val)
{
  pairlist h;

  new(h);
  h->name = newstring(name);
  h->value = newstring(val);
  h->next = *p;
  *p = h;
}


/* print_list_recursive -- print LI items for all entries in list */
static void print_list_recursive(const pairlist list)
{
  conststring url;

  /* ToDo: Escape double quotes */
  if (list) {
    print_list_recursive(list->next);
    url = base ? URL_s_absolutize(base, list->name) : list->name;
    printf("<li><a class=\"%s\" href=\"%s\">%s</a></li>\n",
	   list->value, url, url);
  }
}


/* print_list -- print an OL with the entries of list */
static void print_list(const pairlist list)
{
  if (list) {
    printf("<ol class=\"urllist\">\n");
    print_list_recursive(list);
    printf("</ol>\n");
  }
}


/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_error = true;
}


/* start -- called before the first event is reported */
void* start(void)
{
  return NULL;
}

  
/* end -- called after the last event is reported */
void end(void *clientdata)
{
  /* If we still have a list, print it here */
  if (list) {
    print_list(list);
    pairlist_delete(list);
    list = NULL;
  }
}


/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  printf("<!--%s-->", commenttext);
}


/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  printf("%s", text);
}


/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi, string fpi,
		 string url)
{
  if (fpi && url)
    printf("<!DOCTYPE %s PUBLIC \"%s\" \"%s\">\n", gi, fpi, url);
  else if (fpi)
    printf("<!DOCTYPE %s PUBLIC \"%s\">\n", gi, fpi);
  else if (url)
    printf("<!DOCTYPE %s SYSTEM \"%s\">\n", gi, url);
  else
    printf("<!DOCTYPE %s>\n", gi);
}


/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  printf("<?%s>", pi_text);
}


/* print_attrs -- print attributes */
static void print_attrs(const pairlist attribs)
{
  pairlist p;

  /* ToDo: Distinguish SGML (a NULL value means that the name is the
     value and the actual attribute name is implicit) and XML? */
  for (p = attribs; p; p = p->next)
    printf(" %s=\"%s\"", p->name, p->value ? p->value : p->name);
}


/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  int i;
  conststring url;

  /* Store any URLs from attributes */
  for (i = 0; i < sizeof(attname)/sizeof(*attname); i++)
    if ((url = pairlist_get(attribs, attname[i])))
      pairlist_push(&list, url, attname[i]);

  printf("<%s", name);
  print_attrs(attribs);
  printf(">");
}


/* handle_emptytag -- called after an empty element is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  int i;
  conststring url;

  /* Store any URLs from attributes */
  for (i = 0; i < sizeof(attname)/sizeof(*attname); i++)
    if ((url = pairlist_get(attribs, attname[i])))
      pairlist_push(&list, url, attname[i]);

  printf("<%s", name);
  print_attrs(attribs);
  printf(" />");
}


/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  /* Just before </body>, print the list. or if we see </html> and
     haven't printed the list yet, print it there */
  if (list && (strcasecmp(name,"body") == 0 || strcasecmp(name,"html") == 0)) {
    print_list(list);
    pairlist_delete(list);
    list = NULL;
  }
  printf("</%s>", name);
}


/* usage -- print usage message and exit */
static void usage(string prog)
{
  fprintf(stderr, "Usage: %s [html-file-or-url]\n", prog);
  exit(2);
}


/* main -- main body */
int main(int argc, char *argv[])
{
  int c, status = 200;

  while ((c = getopt(argc, argv, "b:")) != -1)
    switch (c) {
    case 'b': base = optarg; break;
    default: usage(argv[0]);
    }

  if (optind < argc - 1) usage(argv[0]);
  else if (optind > argc - 1 || eq(argv[optind], "-")) yyin = stdin;
  else yyin = fopenurl(argv[optind], "r", &status);

  if (!yyin) {perror(argv[optind]); exit(2);}
  if (status != 200) errexit("%s : %s\n", argv[optind], http_strerror(status));

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  if (yyparse() != 0) {exit(3);}
  return has_error ? 1 : 0;
}
